llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -4140,6 +4303,92 @@ bool AIE2InstructionSelector::select512BitG_AIE_STORE_SRS(
   }
 }
 
+// Returns the instruction where the Reg is defined ignoring G_BITCAST.
+static MachineInstr *stripBitCast(Register Reg, MachineRegisterInfo &MRI) {
+  MachineInstr *MI = getDefIgnoringCopies(Reg, MRI);
+  if (MI->getOpcode() == TargetOpcode::G_BITCAST)
+    return getDefIgnoringCopies(MI->getOperand(1).getReg(), MRI);
+  return MI;
+}
+
+// If two 256-bit SRS are input to 512-bit STORE through CONCAT, these are
+// combined into VST.SRS
+bool AIE2InstructionSelector::selectG_AIE_STORE_CONCAT_SRS(
+    MachineInstr &StoreI, MachineRegisterInfo &MRI) {
+
+  Register ConcatResult = (StoreI.uses().begin())->getReg();
+  MachineInstr *ConcatOp = MRI.getUniqueVRegDef(ConcatResult);
+
+  assert(ConcatOp && "Expected SSA.");
+
+  // Do not try to combine if one of the concat's defs is used by another
+  // instruction between the concat and the store or if there is a other store
+  // between the concat and the store.
+  if (!canDelayMemOp(*ConcatOp, StoreI, MRI))
```

Comment:
Maybe we need a different name here, or different function. ConcatOp is not a mem op and also there is no problem about crossing stores.

Reply:
This checking is blocking some optimizations. See: 

```
---
name:            A
alignment:       16
legalized:       true
regBankSelected: true
body:             |
  bb.1.entry:
    liveins: $bml0, $bml1, $p0, $r0
    %0:accregbank(<8 x s64>) = COPY $bml0
    %1:accregbank(<8 x s64>) = COPY $bml1
    %2:ptrregbank(p0) = COPY $p0
    %3:gprregbank(s32) = G_CONSTANT i32 1
    %5:gprregbank(s32) = G_CONSTANT i32 0
    %4:vregbank(<8 x s32>) = G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.aie2.I256.v8.acc64.srs), %0(<8 x s64>), %5(s32), %3(s32)
    %6:vregbank(<8 x s32>) = G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.aie2.I256.v8.acc64.srs), %1(<8 x s64>), %5(s32), %3(s32)
    %7:vregbank(<16 x s32>) = G_INTRINSIC intrinsic(@llvm.aie2.concat.I512.I256), %4(<8 x s32>), %6(<8 x s32>)
    %8:gprregbank(s32) = G_CONSTANT i32 9
    %9:modregbank(s20) = G_CONSTANT i20 -128
    G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.aie2.set.ctrl.reg), %8:gprregbank(s32), %3:gprregbank(s32)
    %10:ptrregbank(p0) = G_AIE_POSTINC_STORE %7(<16 x s32>), %2(p0), %9 :: (store (<16 x s32>))
...

```

I tested both cases and they were not combined.

I got:
```
    %0:acc512 = COPY $bml0
    %1:acc512 = COPY $bml1
    %2:ep = COPY $p0
    %4:er = MOV_RLC_imm10_pseudo 0
    %14:mss = COPY %4
    %5:ewl = VSRS_S32_S64_mv_w_srs %0, %14, implicit-def dead $srsrs_of, implicit $crsat, implicit $crrnd
    %13:mss = COPY %4
    %6:ewh = VSRS_S32_S64_mv_w_srs %1, %13, implicit-def dead $srsrs_of, implicit $crsat, implicit $crrnd
    %7:vec512 = REG_SEQUENCE %5, %subreg.sub_256_lo, %6, %subreg.sub_256_hi
    $crsat = MOV_scalar_imm10_pseudo 1
    %11:vec256 = COPY %7.sub_256_lo
    %12:vec256 = COPY %7.sub_256_hi
    VST_dmw_sts_w_ag_idx_imm %12, %2, 32 :: (store (<8 x s32>) into unknown-address + 32)
    %10:ep = VST_dmw_sts_w_ag_pstm_nrm_imm %11, %2, -128 :: (store (<8 x s32>), align 64)
```

Instead of:
```
    %0:acc512 = COPY $bml0
    %1:acc512 = COPY $bml1
    %2:ep = COPY $p0
    %4:er = MOV_RLC_imm10_pseudo 0
    $crsat = MOV_scalar_imm10_pseudo 1
    %15:mss = COPY %4
    VST_SRS_S32_S64_ag_idx_imm %2, 32, %1, %15, implicit-def $srsrs_of, implicit $crsat, implicit $crrnd :: (store (<8 x s32>) into unknown-address + 32)
    %14:mss = COPY %4
    %10:ep = VST_SRS_S32_S64_ag_pstm_nrm_imm %2, -128, %0, %14, implicit-def $srsrs_of, implicit $crsat, implicit $crrnd :: (store (<8 x s32>), align 64)
```

The same applies if I use a G_STORE instead of POSTINC.